src/primitives/hrp.rs

```diff
@@ -0,0 +1,180 @@
+// Written by Clark Moody and the rust-bitcoin developers.
+// SPDX-License-Identifier: MIT
+
+//! Provides an `Hrp` type that represents the human-readable part of a bech32 encoded string.
+//!
+//! > The human-readable part, which is intended to convey the type of data, or anything else that
+//! > is relevant to the reader. This part MUST contain 1 to 83 US-ASCII characters, with each
+//! > character having a value in the range [33-126]. HRP validity may be further restricted by
+//! > specific applications.
+//!
+//! ref: [BIP-173](https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#user-content-Bech32)
+
+#[cfg(feature = "alloc")]
+use alloc::borrow::Cow;
+use core::fmt;
+use core::iter::FusedIterator;
+
+/// Maximum length of the human-readable part, as defined by BIP-173.
+const MAX_HRP_LEN: usize = 83;
+
+/// The human-readable part (human readable prefix before the '1' separator).
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
```

Comment:
In e0bfa91dc74888005246ae80b46d0a620827888f:

I think we should do a case-insensitive comparison rather than deriving PartialEq/Eq.

Reply:
And I think we should impl `Ord` similarly.

Reply:
Done!